File: fe2-android/app/src/main/java/com/github/dedis/student20_pop/utility/security/Hash.java
package com.github.dedis.student20_pop.utility.security;

import android.util.Log;
import com.github.dedis.student20_pop.exceptions.PoPException;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Base64;

/** SHA256 Hashing Class */
public class Hash {

  public static final String TAG = Hash.class.getSimpleName();

  /**
   * Hash some objects using SHA256. Concatenate the object's string representation following the
   * protocol's directive. Then hash the obtained string
   *
   * @param strs : the strings to hash
   * @return the hashed data or null if failed to hash
   * @throws IllegalArgumentException if the data is null or empty
   * @throws PoPException if SHA-256 MessageDigest is unavailable
   */
  public static String hash(String... strs) {
    if (strs == null || strs.length == 0) {
      throw new IllegalArgumentException("cannot hash an empty/null array");
    }
    try {
      MessageDigest digest = MessageDigest.getInstance("SHA-256");
      for (String str : strs) {
        if (str == null || str.length() == 0) {
          throw new IllegalArgumentException("cannot hash an empty/null string");
        }
        String length = Integer.toString(str.length());
        digest.update(length.getBytes(StandardCharsets.UTF_8));
        digest.update(str.getBytes(StandardCharsets.UTF_8));
      }
      byte[] digestBuf = digest.digest();
      return Base64.getEncoder().encodeToString(digestBuf);
    } catch (NoSuchAlgorithmException e) {
      Log.e(TAG, "failed to hash", e);
      throw new PoPException("failed to retrieve SHA-256 instance", e);
    }
  }

  private static String esc(String input) {
    return input.replace("\\", "\\\\").replace("\"", "\\\"");
  }
}

File: be2-scala/src/main/scala/ch/epfl/pop/model/objects/Hash.scala
package ch.epfl.pop.model.objects

import java.nio.charset.StandardCharsets
import java.security.MessageDigest

final case class Hash(base64Data: Base64Data) {
  def getBytes: Array[Byte] = base64Data.getBytes

  override def toString: String = base64Data.toString
}

object Hash {

  val messageDigest: MessageDigest = MessageDigest.getInstance("SHA-256")

  /** Create a base64 encoded hash of an array of strings according to the communication protocol
    *
    * @param data
    *   values to be hashed
    * @return
    *   resulting hash
    */
  def fromStrings(data: String*): Hash = Hash.sha256Hash(data.foldLeft("")((acc, s) => acc + s.length + s))

  /** Create a hash of a string
    *
    * @param data
    *   value to be hashed
    * @return
    *   resulting hash
    */
  def sha256Hash(data: String): Hash = Hash(Base64Data.encode(
    messageDigest.digest(data.getBytes(StandardCharsets.UTF_8))
  ))
}
